### PR TITLE
Service node purging of contract-missing nodes

### PR DESCRIPTION
**Pull request by jagerman**
This implements service node purging of service nodes not in the rewards contract by introducing a ServiceNodePurge pseudo-event that is generated within oxend based on contract observations, but unlike existing L2Event transactions, does not map to a contract event (because "x is missing from the contract" is not something that can be initiated from the contract side).

This works by periodically (every 2h, on mainnet) checking the reward contract service node list and using that list to decide on whether nodes have become missing from the network; if they are, service nodes construct a ServiceNodePurge pseudo-event for inclusion in pulse blocks.

This is very much an out-of-the-norm case (such as a major distributed L2 failure, or an unforseeable future case that requires manual intervention), but if it *does* happen it means the Oxen service node list is in a fundamentally difference state from the upstream contract, and needs correction.

The purge implementation here works similarly to L2 event confirmations: it gets mined into a block by a pulse leader, then confirmed or denied by subsequent quorums.  Unlike L2 events, though, these are pseudo-events generated by oxen nodes themselves and so the criteria for confirming a purge is quite difference.

## Purge list details

Oxend nodes periodically (every 2h) refresh a list of reward contract nodes.  This frequency is kept low as this are potentially quite large requests -- perhaps somewhere in the 300-400kB range for getting all mainnet BLS pubkeys at once from a hex-producing L2 RPC provider, and we don't need this check to fire frequently.

The 2h interval is syncronized by L2 height: all nodes request the data at multiples of 28800 (2h of 250ms Arbitrum blocks).  If that fails (which is likely on initial startup as Arb nodes typically only have 30mins of history) then we fall back and request the current height.

Assuming we *don't* have to resort to the fallback, above, then we consider nodes for purging; such a node must meet these criteria:

- we must have a recent list from the contract.  If we miss it for some reason, or the contract is misbehaving and returning an empty list, we don't want to start purging everyone.
- the target must be a currently registered Oxen service node that was registered at least 24h ago (i.e. 720 oxen blocks).  This age limit is so we don't allow purging of brand new nodes (because our contract knowledge could be up to 2h old).

These are the same checks done for pulse confirmations of purge events.

Additionally we only consider proposing (i.e. by adding into our own mempool) if:

- we obtained the list for the correct target L2 height.  If we resorted to the current value via fallback then our list is still okay for confirmations, but we don't want to make new transactions in the mempool that won't deduplicate with what other nodes are pushing for blocks on the correct intervals.
- we must be running in service node mode and have a fully synced oxen chain.
- HF21+

## Other changes

Also fixes a few random issues:
- removed a confusing (and unneeded) lock parameter when fetching chain height
- added a bls pubkey lookup map alongside the x25519 map, and changed around how they work so that we keep them for recently_removed_nodes (and can thus simplify the map-and-then-scan-recent-nodes code), and rewrote the existing bls lookup code to use this instead of doing a full service node list linear search.
- make the existing bls lookup work more like the x25519 lookup (i.e. return null hash instead of throwing for not found), and then unified the names between the two lookups.
- fixed a thread-safety issue in how recently-removed-nodes were being processed.

As well this rearranges the mempool acceptance code a bit to move some of the checks into the "is ready to go" function rather than having some inside and some outside.

### CODE REVIEW

File: src/bls/bls_aggregator.cpp [hunk 1]
```diff
@@ -235,11 +233,64 @@ namespace {
...
+                    logcat,
+                    "Re-derived (via subtraction) BLS aggregate public key {}",
+                    cpp_agg_pubkey.get());
         }
     }
 #endif
 
+    void log_aggregation_result(
+            std::string_view agg_type,
+            [[maybe_unused]] cryptonote::core& core,
```
Comment (javabudd) [inline src/bls/bls_aggregator.cpp]: nit: grouping maybe_unused arguments together could improve readability

Comment (jagerman) [reply src/bls/bls_aggregator.cpp]: Good point, I've updated.

File: src/l2_tracker/l2_tracker.cpp [hunk 2]
```diff
@@ -496,18 +431,249 @@ void L2Tracker::update_logs() {
...
+                                    "Unexpected service node list retrieval result for SN purge "
+                                    "checks failed (twice): {}",
+                                    maybe_result->dump());
+                        }
+                    }
+                },
+
+                "0x{:x}"_format(purge_height));
+    } else {
+        update_in_progress = false;
```
Comment (Doy-lee) [inline src/l2_tracker/l2_tracker.cpp:620]: Need to upgrade into unique lock for this write

**Review by Doy-lee — APPROVED** [review]
Seems generally OK. Some questions about what happens at the fringe when you're in the SNL and someone starts interacting from the contract side. But I've started a discussion on that out of band.

**Comment by Doy-lee** [comment]
Merged in https://github.com/oxen-io/oxen-core/pull/1746

**Pull request closed without merging**